src/Data/Text/Utf8/BoyerMoore/Automaton.hs

```diff
@@ -330,6 +330,6 @@ buildBadCharTable pattern = runST $ do
 -- Helper functions for easily toggling the safety of this module
 
 -- | Read from a lookup table at the specified index.
-indexTable :: UVector.Unbox a => UVector.Vector a -> Int -> a
+indexTable :: Prim a => TypedByteArray a -> Int -> a
 {-# INLINE indexTable #-}
-indexTable = (UVector.!) -- UVector.unsafeIndex
+indexTable = TBA.unsafeIndex
```

Comment:
Was the performance increase you measured caused by the change to this new type, or was it just the removed bounds checks?

And how sure are we that it cannot go wrong here?

Reply:
Good catch! I assumed that the comment meant that `!` is equivalent to `unsafeIndex` but I see the problem now.

Nevertheless the lookup indices can not be influenced directly by the user:

```
badCharSkip = badCharLookup badCharTable (codeUnitAt haystackIndex)
suffixSkip = suffixLookup suffixTable needleIndex
```

`badCharLookup` only ever uses bytes to index the `badCharTable`, which is constructed as:
```
TBA.replicate 256 $ codeUnitIndex patLen
```

`suffixLookup` only ever uses indices that can be used as code unit indices into the needle, and `suffixTable` has exactly that size:
```
TBA.newTypedByteArray $ codeUnitIndex patLen
```

I will however check how big a difference it makes to use a bounds-checked function to access `TypedByteArray`s.